gax/src/main/java/com/google/api/gax/rpc/mtls/MtlsProvider.java

```diff
@@ -0,0 +1,195 @@
+    @Override
+    public String getenv(String name) {
+      return System.getenv(name);
+    }
+  }
+
+  interface ProcessProvider {
+    public Process createProcess(InputStream metadata) throws IOException;
+  }
+
+  static class DefaultProcessProvider implements ProcessProvider {
+    @Override
+    public Process createProcess(InputStream metadata) throws IOException {
+      if (metadata == null) {
+        return null;
+      }
+      List<String> command = extractCertificateProviderCommand(metadata);
+      return new ProcessBuilder(command).start();
+    }
+  }
+
+  private static final String DEFAULT_CONTEXT_AWARE_METADATA_PATH =
+      String.format(
+          "%s/.secureConnect/context_aware_metadata.json", System.getProperty("user.home"));
+
+  private String metadataPath;
+  private EnvironmentProvider envProvider;
+  private ProcessProvider processProvider;
+
+  /**
+   * The policy for mutual TLS endpoint usage. NEVER means always use regular endpoint; ALWAYS means
+   * always use mTLS endpoint; AUTO means auto switch to mTLS endpoint if client certificate exists
+   * and should be used.
+   */
+  public enum MtlsEndpointUsagePolicy {
+    NEVER,
+    AUTO,
+    ALWAYS;
+  }
+
+  @VisibleForTesting
+  MtlsProvider(
+      EnvironmentProvider envProvider, ProcessProvider processProvider, String metadataPath) {
+    this.envProvider = envProvider;
+    this.processProvider = processProvider;
+    this.metadataPath = metadataPath;
+  }
+
+  public MtlsProvider() {
+    this(
+        new SystemEnvironmentProvider(),
+        new DefaultProcessProvider(),
+        DEFAULT_CONTEXT_AWARE_METADATA_PATH);
+  }
+
+  /**
+   * Returns if mutual TLS client certificate should be used. If the value is true, the key store
+   * from {@link #getKeyStore()} will be used to configure mutual TLS transport.
+   */
+  public boolean useMtlsClientCertificate() {
+    String useClientCertificate = envProvider.getenv("GOOGLE_API_USE_CLIENT_CERTIFICATE");
+    return "true".equals(useClientCertificate);
+  }
+
+  /** Returns the current mutual TLS endpoint usage policy. */
+  public MtlsEndpointUsagePolicy getMtlsEndpointUsagePolicy() {
+    String mtlsEndpointUsagePolicy = envProvider.getenv("GOOGLE_API_USE_MTLS_ENDPOINT");
+    if ("never".equals(mtlsEndpointUsagePolicy)) {
+      return MtlsEndpointUsagePolicy.NEVER;
+    } else if ("always".equals(mtlsEndpointUsagePolicy)) {
+      return MtlsEndpointUsagePolicy.ALWAYS;
+    }
+    return MtlsEndpointUsagePolicy.AUTO;
+  }
+
+  /** The mutual TLS key store created with the default client certificate on device. */
+  public KeyStore getKeyStore() throws IOException {
+    try (InputStream stream = new FileInputStream(metadataPath)) {
+      return getKeyStore(stream, processProvider);
+    } catch (FileNotFoundException exception) {
+      // If the metadata file doesn't exist, then there is no key store, just return null.
+      return null;
```

Comment:
optional: per Effective Java, consider returning an empty keystore instead of a null keystore.

Reply:
I think I will just keep the null keystore for now.